testsuite/tests/typing-unboxed-types/test.ml

```diff
@@ -94,7 +94,7 @@ Error: This type cannot be unboxed because
 (* let rec must be rejected *)
 type t10 = A of t10 [@@ocaml.unboxed] [@@value];;
 [%%expect{|
-type t10 = A of t10 [@@value] [@@unboxed]
+type t10 = A of t10 [@@unboxed]
```

Comment:
In these recursive cases, for unboxed types, it might indeed be appropriate to include a `: value` annotation. But I don't think it should depend on what the user wrote. Instead, I would want to print out a canonical syntax for the type.

Looking around for prior art in this space, I thought about the `-unboxed-types` option, which automatically unboxes all types where this is possible. If I have

```ocaml
type t = MkT of int
```

and say `ocamlc -i` on the file, I get

```
type t = MkT of int
```

Good. Now if I say `ocamlc -unboxed-types -i`, I get

```
type t = MkT of int [@@unboxed]
```

No change in source code. But ocaml is printing out the canonical declaration each time.

On the other hand, if I say

```ocaml
type t = MkT of int [@@boxed]
```

and compile with `ocamlc -unboxed-types -i`, I get

```
type t = MkT of int
```

This is still my canonical representation, but it's potentially confusing, because if I change my code to have that output, it would mean something different than what I wrote (given the use of the `-unboxed-types` flag). So maybe it's "canonical representation in the absence of flags". In any case, it doesn't take into account the user's precise syntax.

Happy to discuss further!

Reply:
Sure - I agree canonicity in this sense would be great.  If you have in mind an analysis that will print `[@@value]` only if it's needed, or some overapproximation of that, I'm all for it.

If we have to pick between printing a canonical representation and printing a type that is accurate, I'd certainly choose the latter.  The current behavior for `@@immediate` and `@@immediate64` is to print them just when the user wrote them, and this doesn't seem to cause any issues, despite not being canonical.

Reply:
I forgot to read my own Note when replying. Though canonicity would be great, that's not actually what's implemented. What's implemented is that we print any user-written annotation on an unboxed type except value. So, basically, I think you and I have the same stance here, except for what to do when the layout is `value`. I think e.g. `type loopy = MkL of loopy [@@unboxed]` should get a `value` layout, and then (I believe) the current behavior is correct. I'll add a test about the `loopy` behavior (admitting the possiility of further discussion on design) so we don't forget to address this.

Reply:
Actually, as I was clarifying the note and the implementation, I've changed my mind, to agree with you. Now, for fully abstract types (i.e. no `=` signs) and for any unboxed types, we're just printing what the user wrote (or not).